File: editorial/ruby/prob3-7-6.rb
# 入力
n = gets.to_i
a = gets.split.map(&:to_i)

# 配列の初期化
dp1 = Array.new(n + 1)
dp2 = Array.new(n + 1)
dp1[0] = 0
dp2[0] = 0

# 動的計画法
(1..n).each do |i|
  dp1[i] = dp2[i - 1] + a[i - 1]
  dp2[i] = [dp1[i - 1], dp2[i - 1]].max
end

# 出力
answer = [dp1[n], dp2[n]].max
puts answer

File: math_algorithm/editorial/go/prob3-7-6.go
package main

import (
	"fmt"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func main() {
	var n int
	fmt.Scan(&n)

	a := make([]int, n)
	for i := 0; i < n; i++ {
		fmt.Scan(&a[i])
	}

	dp1 := make([]int, n+1)
	dp2 := make([]int, n+1)
	dp1[0] = 0
	dp2[0] = 0

	for i := 1; i <= n; i++ {
		dp1[i] = dp2[i-1] + a[i-1]
		dp2[i] = max(dp1[i-1], dp2[i-1])
	}

	answer := max(dp1[n], dp2[n])
	fmt.Println(answer)
}
